### PR TITLE
Minimizing Data Leakage in Machine Learning

### PR DESCRIPTION
**Pull request by collins-a**
Article as index.md.
Hero image as hero.jpg.

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [ ] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by collins-a** [comment]
This fixes #803

File: articles/data-leakage/index.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+It is possible to leak data in pre-processing. If the steps mentioned above are applied to an entire dataset, leakage may manifest. This allows the model to learn beyond the training set by also learning the test set, defeating the purpose of having a test set since it is described as unseen data.
+
+#### Same Data for Testing and Validation
+
+It is standard to split data into separate groups such as testing, training, and validation sets. Not all models require a validation set since we can train a model using the training set and use the test set to test it. As standard as this process is, it becomes a cause of data leakage if not done correctly. For example, if we unintentionally expose the model to test data during training, the model will already have access to data, it is not yet allowed to see. This then allows a model to perform better in its predictions regarding the test data. A problem associated with this performance is that during deployment, the same model performs underwhelmingly. Data leakage impedes models from generalizing well after in the real-world, facing new data.
```
Comment (lalith1403) [inline articles/data-leakage/index.md:29]: Interesting point. But, how would the two sets get mixed up? Elaborate on that.

File: articles/data-leakage/index.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+#### Eliminating Duplicates
+
+Predictive models always go through a form of data preparation. Data preparation involves the transforming of raw data into a format that is suitable for modeling. The processes involved in data preparation include data cleaning, feature selection, feature engineering, and dimensionality reduction. A process that is key to the elimination of duplicates is data cleaning. Data cleaning is characterized by fixing problems in messy, erroneous, corrupt, or noisy data. Firstly, these problematic instances are identified. They may then be handled by getting rid of a problematic row or a column. They may also be handled by replacing instances with new values.
+However, we mentioned that one of the causes of data leakage is duplicates. Our focus here would be to eliminate them. In data cleaning, we identify duplicate rows and get rid of them. We eliminate duplicates to prevent them from ending up in both test and training sets concerning data leakage.
+
+### Closing
```
Comment (lalith1403) [inline articles/data-leakage/index.md:61]: Conclusion or Wrapping up is more apt.

File: articles/data-leakage/index.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+### Techniques to Minimize Leakage
+
+Since we now know the possible causes of leakage, we should explore the various ways to minimize or prevent data leakage. Let's go through three effective methods.
+
+#### Normalizing Correctly Before Cross-Validation
+
+It is easy to leak data when preparing data for machine learning. As a result, overfitting occurs in training data. This also impacts the performance of the model on test data. We would be given a flattering picture of performance, but it performs poorly when the model is deployed in the real world.
+For instance, data leakage will occur in a scenario where one carries out normalization of an entire dataset, then evaluates the algorithm's performance using cross-validation. [Normalization](https://developers.google.com/machine-learning/data-prep/transform/normalization#:~:text=%20Normalization%20Techniques%20at%20a%20Glance%20%201,a%20wide%20range%20to%20a%20narrow...%20More%20) is a process that aims to improve the performance of a model by transforming features to be on a similar scale. Data normalization is carried out by dividing the data by its mean.
```
Comment (lalith1403) [inline articles/data-leakage/index.md:46]: There are different types of normalization. The last sentence is true only in a specific kind. There are min-max normalization, mean-subtraction, etc. Remove the last sentence.

File: articles/data-leakage/index.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+I described [cross validation](/engineering-education/evaluating-ml-model-performance/) as the use of two datasets to evaluate an algorithm. Specifically, there is a training set and an independent set that is partitioned from an original dataset. These splits of the dataset are called folds, which are often equal in size. A model being evaluated is trained on all folds except one. So, a model will be trained on three and tested on the remaining fold if there are four-folds. The process is repeated until each fold is used for testing.
+From the example above, a clear mistake is the normalizing of an entire dataset. We should note that normalization needs to be carried out after splitting data into train, test, and validation sets. If normalization is carried out before splitting, it will affect the whole dataset instead of the training set. That is, the mean and standard deviation used for normalization will reflect the whole dataset. As a result, information on test and validation sets sips into the training set.
```
Comment (lalith1403) [inline articles/data-leakage/index.md:48]: It is clear that normalization affects the results. But, how does the order of events change the results? Can you elaborate on that?

**Review by lalith1403 — CHANGES_REQUESTED** [review]
Amazing article. Elaborate on a few points and this should be good to go.

**Comment by collins-a** [comment]
Hello @lalith1403, thanks as always, made the requested changes. Check them out when you can. Cheers and happy holidays!

**Comment by hectorkambow** [comment]
plagiarism check done.

**Review by lalith1403 — APPROVED** [review]

**Pull request merged**